training/replication-rebalancing-fault-tolerance.md

```diff
@@ -0,0 +1,140 @@
+
+Each node in your cluster is an equally suitable SQL gateway to your cluster, but to ensure an even balancing of client requests across these nodes, we'll use a TCP load balancer.
+
+For this lab, we'll install the open-source [HAProxy](http://www.haproxy.org/). If you're on a Mac and using Homebrew, use `brew install haproxy`.
+
+After the installation is complete, in a new terminal, run the [`cockroach gen haproxy`](/stable/generate-cockroachdb-resources.html) command:
+
+{% include copy-clipboard.html %}
+~~~ shell
+$ cockroach gen haproxy --insecure
+~~~
+
+This command generates an `haproxy.cfg` file automatically configured to work with the nodes of your running cluster.
+
+Open `haproxy.cfg`, and change `bind :26257` to `bind :26000`. This changes the port on which HAProxy accepts requests to a port that is not already in use by a node and that won't be used by the nodes you'll add later.
+
+~~~
+global
+  maxconn 4096
+
+defaults
+    mode                tcp
+    timeout connect     10s
+    timeout client      1m
+    timeout server      1m
+
+listen psql
+    bind :26000
+    mode tcp
+    balance roundrobin
+    server cockroach1 localhost:26257
+		...
+~~~
+
+Start HAProxy, with the `-f` flag pointing to the `haproxy.cfg` file:
+
+{% include copy-clipboard.html %}
+~~~ shell
+$ haproxy -f haproxy.cfg
+~~~
+
+### Step 2. Start a Load Generator
+
+Now that you have a load balancer running in front of your cluster, we'll generate a load against it to see how CockroachDB behaves.
+
+To do this, we need to install two things:
+
+- [Go](https://golang.org/dl/). If you're on a Mac and using Homebrew, use `brew install go`.
+- [CockroachDB's version of YCSB](https://github.com/cockroachdb/loadgen/tree/master/ycsb): `go get github.com/cockroachdb/loadgen/ycsb`
+
+After installing both of those, in a new terminal, start `ycsb`, pointing it at HAProxy's port:
+
+{% include copy-clipboard.html %}
+~~~ shell
+$ $HOME/go/bin/ycsb -duration 20m -tolerate-errors -concurrency 2 -rate-limit 100 'postgresql://root@localhost:26000?sslmode=disable'
+~~~
+
+This command initiates 2 concurrent client workloads for 20 minutes, but limits each worker to 100 operations per second (since you're running everything on a single machine).
+
+### Step 3. Watch Data Rebalance
+
+Now open the Admin UI at `http://localhost:8080` and hover over the **SQL Queries** graph at the top. After a minute or so, you'll see that the load generator is executing approximately 95% reads and 5% writes across all nodes:
+
+<img src="{{ 'images/admin_ui_sql_queries.png' | relative_url }}" alt="CockroachDB Admin UI" style="border:1px solid #eee;max-width:100%" />
+
+Scroll down a bit and hover over the **Replicas per Node** graph. You should see a roughly even mix among the nodes in your cluster:
+
+<img src="{{ 'images/admin_ui_replicas_migration.png' | relative_url }}" alt="CockroachDB Admin UI" style="border:1px solid #eee;max-width:100%" />
+
+### Step 4. Kill a Node
```

Comment:
I would use "Shut down" here and below, since `kill` is worse and will actually lead to up to 10s of unavailability. Lab idea: kill a node and try to see that this is worse from the graphs.

Reply:
Done.